---
title: Complexity in biology
author: Gergo Palfalvi
date: '2017-05-16'
slug: complexity-in-biology
categories:
  - post
  - biology
tags:
  - bioinformatics
  - biology
  - network
  - complexity
summary: Nothing.
---

<!-- BLOGDOWN-HEAD -->
<!-- /BLOGDOWN-HEAD -->

<!-- BLOGDOWN-BODY-BEFORE -->
<!-- /BLOGDOWN-BODY-BEFORE -->
<blockquote>
<p><em>I think the next century will be the century of complexity.</em></p>
<p>by Stephen Hawking at 2000. January</p>
</blockquote>
<p>That century is here, we already passed 17 years of it and it looks Hawking was right. We dig into complexity, we generate huge amount of data and easily share everything. Biology is not an exceptation, NGS (New Generation Sequencing) techniques allow us to have mass amount of nucleotide sequence data from different kind of sequencing, most popularly WGS (Whole Genome Sequencing) and RNA-seq. However these generates massive amount of information, most of the biologists tend to still select and focus on one or maximum a few single genes, usually transcription factors (TF). This is somehow understandable, since we are in the middle of a <strong>paradigm shift</strong> where biology will transform from a qualitaty-driven science to a quantitative science. What does it mean? In biology it was (and is) widely accepted to describe a phenomena with only words. Statistics class is already integrated in biology teaching, but what about <strong>bioinformatics</strong>? It is not widely teached and also often treated as a “black sheep”. There is a <a href="http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2002050">nice article</a> in the topic written by <a href="http://www.markowetzlab.org/">Florian Markowetz</a>.</p>
<blockquote>
<p><em>“Pipette biologist. Microscopy biologist. Cell culture biologist. Have you ever heard any of those job titles? No, of course not. All are biologists, because it is the questions you address that matter, not the tools you use, and computational biologists are just biologists using a different tool.”</em></p>
</blockquote>
<p>Anyway, more and more student wants to work with “sexy” tools in biology, including genomic sciences, but universities in general are not ready to prepare them for these things. I was in very similar situation in a few years, our bioinformatics class was exhausted in a few online BLAST search while I was already working with RNA-seq data analysis. For me there is a special field I am love with and want to understand and apply: <strong>Network science</strong> or Systems biology, network biology. And also there is a twist in the story, since I do not work with regular model plants, like Arabidopsis or Rice. For them there are many types of networks available including protein-protein interaction, metabolic or just a simple co-expression. For a non-model or emerging model plant nothing is available and the most easiest things to do are genome sequencing and transcriptomic data. Draw and analyse network from transcriptomic data is not a well-established , highly evolving part of biology, but can open mush more information then a simple differential gene expression analysis. The 4 big steps in gene regulatory network (GRN) inference analysis:</p>
<ol style="list-style-type: decimal">
<li>Experimental design and mass-information production (RNA-seq, sRNA-seq, ChIP-seq etc.)</li>
<li>GRN building</li>
<li>Topological analysis of the network</li>
<li>Analyse network dinamics</li>
</ol>
<div id="experimental-design-and-mass-information-production" class="section level2">
<h2>Experimental design and mass-information production</h2>
<p>It is important to design the sample collection in the aim of your study. If the phenomena remains hidden possibly we still can build a network from the data and analyse it, but the our main interest can be invisible or hidden.<br />
Also we can think the level of the regulation in the GRN. With RNA-seq we can see expressional dependencies between genes (either linear or non-linear relationships). smallRNA-seq allows to incorporate sRNA regulation level, which can add a direct negative link to the target (especially if we have quantitative sRNA-seq data). with Chromatin Immunprecipitation coupled with NGS sequencing (ChIP-seq) we can get direct transcriptional regulatory motifs from DNA-binding proteins (mainly transcriptional factors and histone variants/modifications). There are several other techniques we can expand our dataset even with genetic or epigenetic regulators.</p>
</div>
<div id="grn-building" class="section level2">
<h2>GRN building</h2>
<p>Probably this is the most difficult part, since there is no “gold standard” method and newer and better methods are continously appearing day-by-day. However I list a few methods/packages for R or Python which can be useful if you are thinking to perform some GRN inference.</p>
<p>First of all I can recommend the <a href="http://dreamchallenges.org/">DREAM</a> GRN Inference paper: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3512113/"><em>Wisdom of crowds for robust gene network inference</em></a> and <a href="http://barabasi.com">Albert-Laszlo Barabasi’s webpage</a>.</p>
<p>Interesting packages:</p>
<ul>
<li>igraph</li>
<li>RedeR</li>
<li>WGCNA</li>
<li>Genie3</li>
<li>CLR</li>
<li>ARACNE</li>
<li>MRNET</li>
<li>Infelerator/cMonkey</li>
<li>NIMEFI</li>
<li>Tigress</li>
<li>netter</li>
<li>GRACE</li>
</ul>
</div>
